### PR TITLE
Wire up full redis cache of active usernames

### PR DESCRIPTION
**Pull request by Jeremy-Barnes**
READY TO MERGE.

Some notes: Ignore the tests. They work but that is some truly narsty code I've written.
Lots of moving stuff around making the diffs look bigger than they really are (moving the old SQL UserRepo to UserDbRepo makes git explode), try to ignore that where possible.

Mainly interested in your thoughts on the main repo + the redis caches. I'm caching users like 4 different ways basically as indexing.

Also a new API!

### CODE REVIEW

**Comment by acmmlu** [comment]
Maybe I'm crazy or maybe I smoked a little bit too much python in my youth, but I really like the idea of an attribute-based decorator that marks either a method or a repository class for caching.

Aside from that, what do you think about adding one more layer? A UserRepository class that handles calling the UserCacheRepository/UserDbRepository as needed, and then have UserCacheRepositor just talk to redis like how UserDbRepository just talks to sql

**Comment by Jeremy-Barnes** [comment]
I ALSO LIKE AN ATTRIBUTE but I think all the joins and stuff I am doing with Redis and all of Redis' datatypes might be to complex to wire up entirely with an attribute. Open to ideas though, if you see something I dont. 

What purpose does the additional layer of wrapper class serve?

**Comment by Jeremy-Barnes** [comment]
@acmmlu changed my mind on the wrapper - it seemed like needless abstraction to me, but it occurs to me as I work on the online username list that the way I have it now ties the Repo directly to the implementation. Adding a wrapper allows me to store data in different primary datastores without the implementation details being tied to the contract.

File: Tests/IntegrationTests/AdminTests.cs [hunk 1]
```diff
@@ -10,11 +10,14 @@
...
 using CritterServer.Contract;
+using CritterServer.DataAccess.Db;
+using CritterServer.DataAccess.Caching;
+using StackExchange.Redis;
 
 namespace Tests.IntegrationTests
 {
     /// <summary>
-    /// Creaated once, reused for all tests in AdminTests
+    /// Created once, reused for all tests in AdminTests
```
Comment (Jeremy-Barnes) [inline Tests/IntegrationTests/AdminTests.cs:20]: hey @acmmlu remember june of 2020 when you started your internship and also my keyboard was hot trash?

**Comment by Jeremy-Barnes** [comment]
@acmmlu its done and ready for review!

File: CritterServer/DataAccess/Caching/UserCacheRepository.cs [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+    public interface IUserCacheRepository : IRepository
+    {
+        Task Cache(User user);
+        Task Evict(User user);
+        Task Dump();
+        Task<CacheResponse<int, User>> RetrieveUsersByIds(params int[] userIds);
+        Task<CacheResponse<string, User>> RetrieveUsersByUserNames(params string[] userNames);
+        Task<bool> UserExistsByUserNameOrEmail(string userName, string email);
+        Task UpdateUserCash(int userId, int deltaCash);
+        Task UpdateUsersCash(IEnumerable<(int UserId, int CashDelta)> userIdAndDeltaCash);
```
Comment (acmmlu) [inline CritterServer/DataAccess/Caching/UserCacheRepository.cs:279]: (minor) `params (int UserId, int CashDelta) []` for consistency with other methods?

File: CritterServer/DataAccess/UserRepository.cs [hunk 3]
```diff
@@ -10,168 +13,149 @@ namespace CritterServer.DataAccess
...
+            return await DbRepo.UserExistsByUserNameOrEmail(userName, email);
         }
 
-        public async Task<User> RetrieveDevByUserName(string userName)
+        public async Task<IEnumerable<string>> RetrieveOnlineUsernames(int[] targetUsers)
         {
-            if (userName == null || userName.Length == 0) return null;
+            var user = new User();
+            var users = new ConcurrentBag<string>();
+            await Task.WhenAll(targetUsers.Select(t => CacheRepo.RetrieveFieldForUser(t, nameof(user.UserName)).ContinueWith(u => { if (!string.IsNullOrEmpty(u.Result)) users.Add(u.Result); })));
```
Comment (acmmlu) [inline CritterServer/DataAccess/UserRepository.cs:135]: (minor) break this line up pls.

**Review by acmmlu — APPROVED** [review]
I like this setup for the repository stuff much better.

As for 4-way caching users, how do you feel about only caching userid -> user but also caching mappings for username -> userid, email -> userid? 2x lookups but x/4 memory footprint. I'd say not worth at the moment if it balloons the complexity of the queries. We already have retrieveCachedUserId by username and by email so it feels reasonable.

File: CritterServer/DataAccess/Caching/UserCacheRepository.cs (hunk 2 above)
Comment (Jeremy-Barnes) [reply CritterServer/DataAccess/Caching/UserCacheRepository.cs:279]: FINE. THANKS.

**Pull request merged**